Deployment history

fredmaggiowski opened the pull request:
Added new parameter `deployments` to `get` command and new flag `--project|-p` to select a specific project identifier

`miactl get deployments -p project-id`

davidebianchi commented on sdk/deploy_test.go:
```diff
@@ -0,0 +1,140 @@
+		require.True(t, errors.Is(err, ErrProjectNotFound))
+	})
+
+	t.Run("Error occurs when downloading deploy history", func(t *testing.T) {
```
```suggestion
	t.Run("http error occurs when downloading deploy history", func(t *testing.T) {
```

davidebianchi commented on sdk/deploy_test.go:
```diff
@@ -0,0 +1,140 @@
+
+		history, err := client.GetHistory(DeployHistoryQuery{ProjectID: "project-2"})
+		require.Nil(t, history)
+		require.NotNil(t, err)
```
Here you could use `require.NoError` function. An assertion about the error type or message?

fredmaggiowski replied:
```suggestion
		require.NoError(t, err)
```

davidebianchi commented on sdk/deploy_test.go:
```diff
@@ -0,0 +1,140 @@
+	})
+
+	t.Run("Error occurs when downloading deploy history", func(t *testing.T) {
+		projectsResponseBody := `[{"_id":"mongo-id-1","name":"Project 1","configurationGitPath":"/clients/path","projectId":"project-1","environments":[{"label":"Development","value":"development","cluster":{"hostname":"127.0.0.1","namespace":"project-1-dev"}}],"pipelines":{"type":"gitlab"}},{"_id":"mongo-id-2","name":"Project 2","configurationGitPath":"/clients/path/configuration","projectId":"project ...
```
The valid projects response could be imported from a test file, how do you think?
In those tests it's repeated some times. We could create a `testdata` folder and put it here

davidebianchi commented on sdk/deploy_test.go:
```diff
@@ -0,0 +1,140 @@
+		require.NotNil(t, err)
+	})
+
+	t.Run("Error on malformed history items", func(t *testing.T) {
```
Also here, write why it's malformed history items

fredmaggiowski replied:
It's in the error assertion but OK

```suggestion
	t.Run("Error on malformed history items (invalid DeployItem.ID)", func(t *testing.T) {
```

davidebianchi commented on sdk/deploy_test.go:
```diff
@@ -0,0 +1,140 @@
+
+		history, err := client.GetHistory(DeployHistoryQuery{ProjectID: "project-2"})
+		require.Nil(t, history)
+		require.NotNil(t, err)
```
`require.NoError`

fredmaggiowski replied:
```suggestion
		require.NoError(t, err)
```

davidebianchi commented on sdk/deploy_test.go:
```diff
@@ -0,0 +1,140 @@
+
+		history, err := client.GetHistory(DeployHistoryQuery{ProjectID: "project-2"})
+		require.Nil(t, err)
+		require.Equal(t, 3, len(history))
```
Here you test the correct length of the response but it's never tested that history struct is correctly filled with all json field

fredmaggiowski commented on cmd/get_test.go:
```diff
@@ -143,3 +220,14 @@ func assertMockProjectsCorrectlyRendered(t *testing.T, rows []string) {
+
+func assertMockDeploymentsCorrectlyRendered(t *testing.T, rows []string) {
+	expectedHeaders := "# | STATUS | DEPLOY TYPE | DEPLOY REF | MADE BY | DURATION | FINISHED | VIEW LOG"
+	expectedRow1 := "123 | running | deploy_all | v1.2.3 | John Smith | 12s | 2020-01-12 22:33:44.000000012 +0000 UTC | https://web.url/"
```
The format is RFC3339 which is ISO 8081 compliant. We could improve date display by using a much more human readable format but I strongly advice, for coherence, that it should be same in the whole application.

Check out this list of available format: https://pkg.go.dev/time?tab=doc#pkg-constants 

Apparently the default used is RFC3339Nano, surely we can remove those nanoseconds :)

davidebianchi replied:
RFC822? But it will be better with entire year instead of only the last two digits

fredmaggiowski replied:
I've implemented a custom format and placed it in renderer package, you'll see it in a while

fredmaggiowski commented on cmd/get.go:
```diff
@@ -16,6 +22,19 @@ func newGetCmd() *cobra.Command {
+			switch args[0] {
+			case "projects", "project":
+			case "deployment", "deployments":
+				if err := cmd.MarkFlagRequired("project"); err != nil {
```
This allows us to mark the flag as required only if we've provided the `deployment(s)` argument. There should be tests.

If the project ID is not set the RunE function fails (I can remove the manual check performed right beneath at line 31 which is redundant.

As for the error yes it fails if project flag hasn't been previously registered which could never happen

fredmaggiowski commented on sdk/client.go:
```diff
@@ -32,9 +43,12 @@ var (
 )
 
-// New returns the MiaSdkClient to be used to communicate to Mia Platform
+// New returns the MiaSdkClient to be used to communicate to Mia Platform.
```
Why the suggestion? Function, package and structure documentations should end with a dot.

davidebianchi replied:
I think in the end of the statement, not in the middle :)

fredmaggiowski replied:
ops

fredmaggiowski commented on sdk/deploy.go:
```diff
@@ -0,0 +1,89 @@
+
+// GetHistory interacts with Mia Platform APIs to retrieve a list of the lastest deploy.
+func (d DeployClient) GetHistory(query DeployHistoryQuery) ([]DeployItem, error) {
+	req, err := d.JSONClient.NewRequest(http.MethodGet, "api/backend/projects/", nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var projects Projects
+	if _, err := d.JSONClient.Do(req, &projects); err != nil {
+		var httpErr *jsonclient.HTTPError
+		if errors.As(err, &httpErr) {
+			return nil, httpErr
+		}
+		return nil, fmt.Errorf("%w: %s", ErrGeneric, err)
+	}
+
+	var project *Project
+	for _, p := range projects {
+		if p.ProjectID == query.ProjectID {
+			project = &p
+			break
+		}
+	}
+	if project == nil {
+		return nil, fmt.Errorf("%w: %s", ErrProjectNotFound, query.ProjectID)
+	}
```
Ok, I'll refactor it

davidebianchi commented on sdk/client.go:
```diff
@@ -32,9 +43,12 @@ var (
 
-// New returns the MiaSdkClient to be used to communicate to Mia Platform
+// New returns the MiaSdkClient to be used to communicate to Mia Platform.
 // Console api
```
```suggestion
// Console api.
```

fredmaggiowski replied:
```suggestion
// Console api.
```

fredmaggiowski commented on sdk/deploy_test.go:
```diff
@@ -0,0 +1,140 @@
+	})
+
+	t.Run("Error occurs when projects fetch holds malformed data", func(t *testing.T) {
+		projectsResponseBody := `[{"_id":9876543,"name":"Project 1","configurationGitPath":"/clients/path","projectId":"project-1","environments":[{"label":"Development","value":"development","cluster":{"hostname":"127.0.0.1","namespace":"project-1-dev"}}],"pipelines":{"type":"gitlab"}},{"_id":"mongo-id-2","name":"Project 2","configurationGitPath":"/clients/path/configuration","projectId":"project-2"," ...
```
Yes it's that, I think these tests are now redundant since they are related to the getProjectByID, i'm removing them

fredmaggiowski commented on sdk/deploy.go:
```diff
@@ -0,0 +1,89 @@
+}
+
+// GetHistory interacts with Mia Platform APIs to retrieve a list of the lastest deploy.
+func (d DeployClient) GetHistory(query DeployHistoryQuery) ([]DeployItem, error) {
```
Slices are already passed by reference, this is a copy of the reference, not the slice itself so.

fredmaggiowski replied:
I don't see the advantages of using a pointer in this context, by passing a copy we ensure that the query can not be altered by the function itself.

fredmaggiowski commented:
Finally had the time to fix the PR, I've added a custom date formatter in order to allow usage of the same format throughout the codebase; added a sdk utility to read test data for files, I guess now it should be ready to be merged :)

davidebianchi commented on sdk/deploy_test.go:
```diff
@@ -0,0 +1,188 @@
+		require.Equal(t, &http.Cookie{Name: "sid", Value: "my-random-sid"}, cookieSid)
+	}
+
+	historyRequestAsserions := func(t *testing.T, req *http.Request) {
```
`historyRequestAsserions` --> `historyRequestAssertions `

fredmaggiowski replied:
fixed typo

davidebianchi commented on sdk/projects.go:
```diff
@@ -63,3 +64,33 @@ func (p ProjectsClient) Get() (Projects, error) {
+	var projects Projects
+	if _, err := client.Do(req, &projects); err != nil {
+		var httpErr *jsonclient.HTTPError
+		fmt.Printf(">>>> %s - %s\n", err.Error(), reflect.TypeOf(err))
```
Remove this print

fredmaggiowski replied:
removed

davidebianchi commented on sdk/projects_test.go:
```diff
@@ -102,34 +100,95 @@ func TestProjectsGet(t *testing.T) {
-		JSONClient: client,
-	}
-}
+	t.Run("Network error occurs during projectId fetch", func(t *testing.T) {
```
Why network error? This is an http error, not a netword error

fredmaggiowski replied:
what do you mean? I can specify it's an HTTP error but HTTP errors are network errors 🤔

fredmaggiowski replied:
Anyhow since it's a test case name I've made it much more explicit :)

davidebianchi commented on sdk/utiltities_test.go:
```diff
@@ -0,0 +1,72 @@
+package sdk
```
Wrong name of the file, `utiltities` --> `utilities`

fredmaggiowski replied:
yeah, renamed it
